cardano-cli/test/cardano-cli-test-lib/Test/Cardano/CLI/Polysemy.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Test.Cardano.CLI.Polysemy
+  ( cardanoCliPath
+  , execCardanoCli
+  , execCardanoCli_
+  , localWorkspace
+  ) where
+
+import           HaskellWorks.Polysemy
+import           HaskellWorks.Polysemy.Error.Types
+import           HaskellWorks.Polysemy.Hedgehog
+import           HaskellWorks.Polysemy.Hedgehog.Assert
+import           HaskellWorks.Polysemy.Hedgehog.Process
+import           HaskellWorks.Polysemy.Prelude
+
+cardanoCliPath :: FilePath
+cardanoCliPath = "cardano-cli"
+
+-- | Execute cardano-cli via the command line.
+--
+-- Waits for the process to finish and returns the stdout.
+execCardanoCli :: ()
+  => HasCallStack
+  => Member (Embed IO) r
+  => Member Hedgehog r
+  => Member Log r
+  => [String]
+  -- ^ Arguments to the CLI command
+  -> Sem r String
+  -- ^ Captured stdout
+execCardanoCli args = withFrozenCallStack $
+  execFlexOk "cardano-cli" "CARDANO_CLI" args
+    & trapFail @GenericError
+    & trapFail @IOException
+
+execCardanoCli_ :: ()
+  => HasCallStack
+  => Member (Embed IO) r
+  => Member Hedgehog r
+  => Member Log r
```

Comment:
We can log in test and non-test functions and those will get annotated in the test report.

Reply:
Where does this Log effect come from?

Reply:
where does this Log effect come from?

Reply:
The `Log` effect comes from `polysemy-log`.

Reply:
The interpreters for the `Log` effect are registered by `propertyOnce` here:

https://github.com/haskell-works/hw-polysemy/blob/main/hedgehog/HaskellWorks/Polysemy/Hedgehog/Property.hs

See the lines:

```
  & interpretLogDataLog
  & setLogLevel (Just Info)
  & interpretDataLogHedgehog formatLogEntry getLogEntryCallStack
  & interpretDataLogHedgehog id (const GHC.callStack)
  & interpretTimeGhc
```